src/odemis/odemisd/main.py

```diff
@@ -321,78 +322,96 @@ def _terminate_all_alive(self):
-
-            for c in parent_less:
-                cname = c.name
-                logging.debug("Stopping comp %s", cname)
-                # TODO: update the .alive VA every time a component is stopped?
-                # maybe not necessary as we are finishing _everything_
+        # alive component -> components which depends on it
+        self._dependents = {c: set() for c in set(mic.alive.value)}  # comp -> set of comps (all the comps that depend on it)
+        # Children will be terminated when their parents are terminated, so no need to keep them in .dependents
+        for comp in list(self._dependents.keys()):  # ._dependents will be modified, so we need a copy
+            for child in comp.children.value:
+                self._dependents.pop(child, None)
+        for comp in self._dependents:
+            deps = comp.dependencies.value
+            for dep in deps:
+                # if a component was created by delegation, use its creator
+                # instead of itself as "used" by the other components, to ensure
+                # they might not be terminated too early.
                 try:
-                    # FIXME: don't try to terminate the children created by
-                    # delegation after terminating their parent (and container)
-                    # For now it just causes a warning message, but it'd be
-                    # better to not have the warning at all.
-                    c.terminate()
+                    d = dep.parent.value if dep.parent else dep
+                    self._dependents[d].add(comp)
                 except Exception:
-                    logging.warning("Failed to terminate component '%s'", cname, exc_info=True)
-
-                # Terminate the container if that was the component for which it
-                # was created.
-                # TODO: check there is really no component still running in the
-                # container?
-                if cname in self._instantiator.sub_containers:
-                    container = self._instantiator.sub_containers[cname]
-                    logging.debug("Stopping container %s", container)
-                    try:
-                        container.terminate()
-                    except Exception:
-                        logging.warning("Failed to terminate container %r", container, exc_info=True)
-                    del self._instantiator.sub_containers[cname]
-
-                # remove from the graph
-                del parents[c]
-                for p in parents.values():
-                    p.discard(c)
-                alive.discard(c)
-                try:
-                    mic.alive.value = alive
-                except Exception:
-                    logging.warning("Failed to update the alive VA", exc_info=True)
+                    # if the dependency has been terminated already, dep.parent will be None, so trying
+                    # to get the value will result in an AttributeError
```

Comment:
That comment doesn't sound correct. If `dep.parent` is None, it's evaluated as false, so the `.value` is never attempted.
Please check more what's going on. Also, you can mention that this code is mostly to handle components which died "early" due to accidents (eg, segfaults).